components/service/sync-autofill/src/test/java/mozilla/components/service/sync/autofill/DefaultCreditCardValidationDelegateTest.kt

```diff
@@ -0,0 +1,145 @@
+        runBlocking {
+            val creditCardFields = UpdatableCreditCardFields(
+                billingName = "Pineapple Orange",
+                cardNumber = "4111111111111111",
+                expiryMonth = 2,
+                expiryYear = 2028,
+                cardType = "visa"
+            )
+            val creditCard = storage.addCreditCard(creditCardFields)
+            val result = validationDelegate.validate(creditCard).await()
+
+            assertEquals(Result.CanBeUpdated(creditCard), result)
+        }
+
+    @Test
+    fun `WHEN existing credit card matches by guid only, THEN update the credit card in storage`() =
+        runBlocking {
+            val creditCardFields = UpdatableCreditCardFields(
+                billingName = "Pineapple Orange",
+                cardNumber = "4111111111111113",
+                expiryMonth = 2,
+                expiryYear = 2028,
+                cardType = "visa"
+            )
+            val creditCard = storage.addCreditCard(creditCardFields)
+            val newCreditCard = createCreditCard(guid = creditCard.guid)
+            val result = validationDelegate.validate(newCreditCard).await()
+
+            assertEquals(Result.CanBeUpdated(creditCard), result)
+        }
+
+    @Test
+    fun `WHEN existing credit card matches by card number only, THEN update the credit card in storage`() =
+        runBlocking {
+            val creditCardFields = UpdatableCreditCardFields(
+                billingName = "Pineapple Orange",
+                cardNumber = "4111111111111114",
+                expiryMonth = 2,
+                expiryYear = 2028,
+                cardType = "visa"
+            )
+            val creditCard = storage.addCreditCard(creditCardFields)
+            val newCreditCard = createCreditCard(
+                guid = creditCard.guid + 1,
+                billingName = creditCard.billingName,
+                cardNumber = creditCard.cardNumber,
+                expiryMonth = creditCard.expiryMonth,
+                expiryYear = creditCard.expiryYear,
+                cardType = creditCard.cardType
+            )
+            val result = validationDelegate.validate(newCreditCard).await()
+
+            assertEquals(Result.CanBeUpdated(creditCard), result)
+        }
+
+    @Test
+    fun `WHEN existing credit card does not match by guid and card number, THEN add the new credit card to storage`() =
+        runBlocking {
+            val newCreditCard = createCreditCard(guid = "2")
+            val creditCardFields = UpdatableCreditCardFields(
+                billingName = "Pineapple Orange",
+                cardNumber = "4111111111111116",
+                expiryMonth = 2,
+                expiryYear = 2028,
+                cardType = "visa"
+            )
+            storage.addCreditCard(creditCardFields)
+
+            val result = validationDelegate.validate(newCreditCard).await()
+
+            assertEquals(Result.CanBeCreated, result)
+        }
+}
```

Comment:
Note how your tests assume that weird/bad things won't happen (they don't test for them). See my comment in the validate method.